sqlglot/parser.py

```diff
@@ -3267,8 +3267,10 @@ def _parse_join(
             kwargs["on"] = self._parse_assignment()
         elif self._match(TokenType.USING):
             kwargs["using"] = self._parse_using_identifiers()
-        elif not isinstance(kwargs["this"], exp.Unnest) and not (
-            kind and kind.token_type == TokenType.CROSS
+        elif (
+            not (outer_apply or cross_apply)
+            and not isinstance(kwargs["this"], exp.Unnest)
+            and not (kind and kind.token_type == TokenType.CROSS)
         ):
             index = self._index
             joins: t.Optional[list] = list(self._parse_joins())
```

Comment:
We were repeatedly calling `_parse_joins` here and then backtracking a few lines below, hence the exponential time increase. I think this section is only important when we have queries that look like:

```sql
SELECT * FROM a JOIN b JOIN c JOIN d ON cond1 ON cond2 ON cond3
```

It looks like T-SQL doesn't expect a join condition for `OUTER/CROSS APPLY`, so I think we can safely avoid this branch:

<img width="701" alt="Screenshot 2024-08-13 at 4 09 03 PM" src="https://github.com/user-attachments/assets/575cd53d-a8ca-4c08-8487-7a4873b188f1">